File: internal/api/models/metrics.go
// Package models provides API request and response types.
package models

import (
	"time"

	"github.com/google/uuid"
)

// PipelineMetrics represents the current metrics for a pipeline.
type PipelineMetrics struct {
	// PipelineID is the ID of the pipeline.
	PipelineID uuid.UUID `json:"pipeline_id"`

	// Status is the current pipeline status.
	Status PipelineStatus `json:"status"`

	// EventsProcessed is the total number of events processed.
	EventsProcessed int64 `json:"events_processed"`

	// EventsPerSecond is the current rate of events being processed.
	EventsPerSecond float64 `json:"events_per_second"`

	// LagSeconds is the current replication lag in seconds.
	LagSeconds float64 `json:"lag_seconds"`

	// LagP95Seconds is the 95th percentile replication lag.
	LagP95Seconds float64 `json:"lag_p95_seconds"`

	// BufferDepth is the number of unprocessed events in the buffer.
	BufferDepth int64 `json:"buffer_depth"`

	// ErrorCount is the total number of errors encountered.
	ErrorCount int64 `json:"error_count"`

	// IcebergCommits is the total number of Iceberg commits made.
	IcebergCommits int64 `json:"iceberg_commits"`

	// IcebergBytesWritten is the total bytes written to Iceberg.
	IcebergBytesWritten int64 `json:"iceberg_bytes_written"`

	// LastEventAt is the timestamp of the last processed event.
	LastEventAt *time.Time `json:"last_event_at,omitempty"`

	// Uptime is the duration since the pipeline started.
	Uptime string `json:"uptime,omitempty"`

	// Tables contains per-table metrics.
	Tables []TableMetrics `json:"tables,omitempty"`
}

// TableMetrics represents metrics for a specific table.
type TableMetrics struct {
	// Schema is the source schema name.
	Schema string `json:"schema"`

	// Table is the source table name.
	Table string `json:"table"`

	// EventsProcessed is the total events processed for this table.
	EventsProcessed int64 `json:"events_processed"`

	// LagSeconds is the current replication lag for this table.
	LagSeconds float64 `json:"lag_seconds"`

	// LastEventAt is the timestamp of the last event for this table.
	LastEventAt *time.Time `json:"last_event_at,omitempty"`
}

// MetricsHistory represents historical metrics data for a pipeline.
type MetricsHistory struct {
	// PipelineID is the ID of the pipeline.
	PipelineID string `json:"pipeline_id"`

	// TimeRange is the time range of the data (e.g., "1h", "24h").
	TimeRange string `json:"time_range"`

	// DataPoints contains the time-series data.
	DataPoints []MetricsDataPoint `json:"data_points"`
}

// MetricsDataPoint represents a single point in the metrics time series.
type MetricsDataPoint struct {
	// Timestamp is the time of this data point.
	Timestamp time.Time `json:"timestamp"`

	// EventsPerSecond is the events per second at this point.
	EventsPerSecond float64 `json:"events_per_second"`

	// LagSeconds is the replication lag at this point.
	LagSeconds float64 `json:"lag_seconds"`

	// BufferDepth is the buffer depth at this point.
	BufferDepth int64 `json:"buffer_depth"`

	// ErrorCount is the cumulative error count at this point.
	ErrorCount int64 `json:"error_count"`
}

// PipelineMetricsResponse wraps pipeline metrics for API responses.
type PipelineMetricsResponse struct {
	Metrics *PipelineMetrics `json:"metrics"`
}

// MetricsHistoryResponse wraps metrics history for API responses.
type MetricsHistoryResponse struct {
	History *MetricsHistory `json:"history"`
}

File: dagster-philotes/dagster_philotes/models/metrics.py
"""Metrics models for Philotes API."""

from datetime import datetime

from pydantic import BaseModel, Field

from dagster_philotes.models.pipeline import PipelineStatus


class TableMetrics(BaseModel):
    """Metrics for a single table in a pipeline."""

    # Note: Using 'schema' directly as field name (valid in Pydantic v2)
    # This matches the JSON key from the Philotes API response
    schema: str = Field(description="Database schema name")
    table: str = Field(description="Table name")
    events_processed: int = Field(default=0, description="Total events processed")
    lag_seconds: float = Field(default=0.0, description="Replication lag in seconds")
    last_event_at: datetime | None = Field(
        default=None, description="Timestamp of last event"
    )

    @property
    def full_name(self) -> str:
        """Get fully qualified table name."""
        return f"{self.schema}.{self.table}"


class PipelineMetrics(BaseModel):
    """Metrics for a Philotes pipeline."""

    pipeline_id: str = Field(description="Pipeline ID")
    status: PipelineStatus = Field(description="Current pipeline status")
    events_processed: int = Field(default=0, description="Total events processed")
    events_per_second: float = Field(default=0.0, description="Current events/sec rate")
    lag_seconds: float = Field(default=0.0, description="Current replication lag")
    lag_p95_seconds: float = Field(default=0.0, description="P95 replication lag")
    buffer_depth: int = Field(default=0, description="Events in buffer")
    error_count: int = Field(default=0, description="Total error count")
    iceberg_commits: int = Field(default=0, description="Iceberg commits made")
    iceberg_bytes_written: int = Field(default=0, description="Bytes written to Iceberg")
    last_event_at: datetime | None = Field(
        default=None, description="Timestamp of last event"
    )
    uptime: str = Field(default="0s", description="Pipeline uptime")
    tables: list[TableMetrics] = Field(
        default_factory=list, description="Per-table metrics"
    )

    @property
    def is_caught_up(self) -> bool:
        """Check if pipeline has caught up (lag < 1 second)."""
        return self.lag_seconds < 1.0

    def is_within_lag(self, max_lag_seconds: float) -> bool:
        """Check if lag is within acceptable threshold."""
        return self.lag_seconds <= max_lag_seconds

    def to_dagster_metadata(self) -> dict[str, float | int | str]:
        """Convert to Dagster metadata dictionary."""
        return {
            "status": self.status.value,
            "events_processed": self.events_processed,
            "events_per_second": self.events_per_second,
            "lag_seconds": self.lag_seconds,
            "buffer_depth": self.buffer_depth,
            "error_count": self.error_count,
            "iceberg_commits": self.iceberg_commits,
            "uptime": self.uptime,
        }
